### PR TITLE
Make IsRecordingOrReplaying calls controllable

### PR DESCRIPTION
**Pull request by kannanvijayan**
Provide a way of controlling which IsRecordingOrReplaying calls return true when recording.

Currently the mechanism exposed is an environment variable containing a path to a file that lists all the calls to be "excluded" from real handling (e.g. always return false).

### CODE REVIEW

**Comment by kannanvijayan** [comment]
Not ready yet, still a bunch of callsites to add labels to.

**Comment by kannanvijayan** [comment]
Still getting this to build, but it's pretty close.  Ran into c++ linkage issues.  Keeping this relatively performant and preventing it from obscuring performance results means we want to pre-parse the excluded set and get it in-memory at startup and then have fast set-membership tests for each `IsRecordingOrReplaying` label.

All the call sites are annotated with labels.  This includes call-sites within `RELEASE_ASSERT` calls, which are kind of uninteresting (won't affect perf otherwise), but decided to label them anyway since there are not many cases like that.

Here are the full list of labels across the codebase:
UniqueID
FromUniqueID
WatchedByDevTools
PokeFullGC
EnsureGCRunner
EnsureCCRunner
ChromeUtils
CCRunnerFired
CanvasRenderingContextHelper::GetContext
CreateAndInitGL
MediaLoadListener::OnStartRequest
RecvUpdateDimensions
RecvRealMouseMoveEvent
RecvRealMouseButtonEvent
RecvRealKeyEvent
OnLocationChange
HangMonitorChild::InterruptCallback
GetCubebContextUnlocked
DecoderFactory::DoCreateDecoder
InstallSoftRealTimeLimitHandler
MediaEncoder::CreateEncoder
CanCreateVP9Decoder
RegisterSupplementalVP9Decoder
CanCreateMFTDecoder
AudioContext::Constructor
AudioContext::Constructor2
PeerConnectionImpl::CreateTransceiverImpl
ScriptLoader::StartLoad
InterruptCallback
Compositor::IsValid
BasicCompositor::BeginFrameForWindow
BasicCompositor::EndRemoteDrawing
ClientLayerManager::AreComponentAlphaLayersEnabled
TextureClient::InitIPDLActor-Forwarder
TextureClient::InitIPDLActor-Compositor
LayerManagerComposite::UpdateAndRender
TiledLayerBufferComposite::UseTiles
TiledLayerBufferComposite::UseTiles
TiledLayerBufferComposite::UseTiles
CompositableParentManager::ReceiveCompositableUpdate
CompositableParentManager::ReceiveCompositableUpdate
CompositableParentManager::ReceiveCompositableUpdate
CompositableParentManager::ReceiveCompositableUpdate
CompositableParentManager::ReceiveCompositableUpdate
CompositorBridgeParent::CompositeToTarget
CompositorBridgeParent::CompositeToTarget
CompositorBridgeParent::SetLayerManager
CompositorManagerChild::CreateSameProcessWidgetCompositorBridge
CompositorManagerParent::CreateSameProcess
CompositorManagerParent::CreateSameProcessWidgetCompositorBridge
CompositorVsyncScheduler::CompositorVsyncScheduler
LayerTransactionParent::IsSameProcess
ShadowLayerForwarder::EndTransaction
ShadowLayerForwarder::ReleaseLayer
ShadowLayerForwarder::Connect
ShadowLayerForwarder::ReleaseCompositable
gfxPlatform::Init
gfxPlatform::InitLayersIPC
gfxPlatform::InitLayersIPC
gfxPlatform::ShutdownLayersIPC
gfxPlatform::InitCompositorAccelerationPrefs
gfxPlatform::InitCompositorAccelerationPrefs
gfxPlatform::InitWebRenderConfig
gfxPlatform::ForceSoftwareVsync
gfxPlatform::ReInitFrameRate
gfxPlatform::GetHardwareVsync
RecordReplayInitializeSurfaceCacheMutext
mojo::core::ports::GetPortId
MessagePump::Run
AutoRecordReplayAssertMessageContents::AutoRecordReplayAssertMessageContents
AutoRecordReplayAssertMessageContents::~AutoRecordReplayAssertMessageContents
ShouldRecordReplayAssertMessageContents
JSDebugger::AddClass
js::Array::MatchNumericComparator
js::EvalKernel
JS::AutoDebuggerJobQueueInterruption::init
JS::AutoDebuggerJobQueueInterruption::runJobs
DebugAPI::onExceptionUnwind
Debugger::recordReplayProcessKind
MemoryProtectionExceptionHandler::isDisabled
frontend::SourceAwareCompiler<Unit>::createSourceAndParser
BytecodeEmitter::BytecodeEmitter.1
BytecodeEmitter::BytecodeEmitter.2
BytecodeEmitter::shouldEmitInstrumentation
FullParseHandler::disableWarnings
GCRuntime::isCompactingGCEnabled
JS::IsIncrementalGCEnabled
js::gc::MapAlignedPages
js::gc::GetPageFaultCount
js::Nursery::init
js::Nursery::enable
BaselineCodeGen<Handler>::emit_ExecutionProgress
BaselineInterpreterGenerator::generate
CodeGenerator::visitInterruptCheck
ExecutablePool::alloc
ReserveProcessExecutableMemory
JS_NewContext
StructuredSpewer::ensureInitializationAttempted
StructuredSpewer::enabled
StructuredSpewer::StructuredSpewer
js::DateTimeInfo::internalResetTimeZone
js::DateTimeInfo::DateTimeInfo
DebugEnvironments::takeFrameSnapshot
js::NewTimeWarpTarget
GlobalHelperThreadState::ensureInitialized
js::StartOffThreadPromiseHelperTask
JS::detail::InitWithFailureDiagnostic
HelperThread::ensureRegisteredWithProfiler
js::ExecuteKernel
js::CheckForInterrupt
js::XDRScript
js::MaybeNoteContentParse
js::CanDoOffThread
js::RecordReplayAssertValue
SharedArrayObject::New
SharedArrayRawBuffer::SharedArrayRawBuffer
TransferableObjectHasher::hash
ReadTimestampCounter
ReadTimestampCounter
ReadTimestampCounter
XDRState::codeScript
DebugState::toggleDebugTrap
wasm::StreamingCompilationAvailable
wasm::EnsureStreamSupport
XPCJSContext::RecordScriptActivity
XPCJSContext::InterruptCallback
XPCJSContext::GetWindowsStackSize
XPCJSRuntime::GCSliceCallback
PresShell::DecApproximateVisibleCount
nsRefreshDriver::Tick
GlobalStyleSheetCache::SetSharedMemory
Loader::ParseSheet
LoaderObserver::OnEndDllLoad
LoaderObserver::Forward
BackgroundHangMonitor::Startup
FinalizationWitnessService::Make
TelemetryImpl::SeteCanRecordBase
TelemetryImpl::SetCanRecordExtend
TelemetryImpl::CreateTelemetryInstance
JSControl::AddRecordingOperation
profiler_start
RefreshTimerVsyncDispatcher::RefreshTimerVsyncDispatcher
RefreshTimerVsyncDispatcher::~RefreshTimerVsyncDispatcher
GfxInfo::Init
CycleCollectedJSRuntime::GCSliceCallback
IncrementalFinalizeRunnable::ReleaseNow
MemoryMapping::GetMemoryMappings
nsTraceRefcnt::DoInitTraceLog
PLDHashTable::ComputeKeyHash
IdlePeriodState::ShouldGetIdleToken
~MozPromise
SharedThreadPool::Release
ThreadEventQueue::GetEvent
ThrottledEventQueue::ExecuteRunnable
nsThread::DoMainThreadSpecificProcessing

**Comment by kannanvijayan** [comment]
Ok it builds now.

File: mfbt/RecordReplay.cpp [hunk 1]
```diff
@@ -157,6 +158,51 @@ void Initialize(int* aArgc, char*** aArgv) {
...
+    // Zero-out any newline chars so the string terminates at end of line.
+    for (int i = 0; i < 128; i++) {
+      if (buf[i] == '\r' || buf[i] == '\n') {
+        buf[i] = 0;
+      }
+    }
+    aExcluded->insert(std::string(buf));
+  }
+}
+static void InitExcludedChecks() {
```
Comment (bhackett1024) [inline mfbt/RecordReplay.cpp:184]: This doesn't seem to be called from anywhere?

**Review by bhackett1024 — APPROVED** [review]

File: mfbt/RecordReplay.cpp (hunk 1 above)
Comment (kannanvijayan) [reply mfbt/RecordReplay.cpp:184]: Good catch.

**Comment by kannanvijayan** [comment]
This is passing macos and windows tests.  The linux failures seem to be spurious.  Gonna land.

**Pull request merged**